katdal/chunkstore_s3.py

```diff
@@ -0,0 +1,129 @@
+    connections.
+
+    The full identifier of each chunk (the "chunk name") is given by
+
+      "<bucket>/<path>/<idx>"
+
+    where "<bucket>" refers to the relevant S3 bucket, "<bucket>/<path>" is
+    the name of the parent array of the chunk and "<idx>" is the index string
+    of each chunk (e.g. "00001_00512"). The corresponding S3 key string of
+    a chunk is therefore "<path>/<idx>".
+
+    Parameters
+    ----------
+    client : :class:`botocore.client.S3` object
+        Pre-configured botocore S3 client
+
+    Raises
+    ------
+    ImportError
+        If botocore is not installed (it's an optional dependency otherwise)
+    """
+
+    def __init__(self, client):
+        if not botocore:
+            raise ImportError('Please install botocore for katdal S3 support')
+        error_map = {EndpointConnectionError: StoreUnavailable,
+                     client.exceptions.NoSuchKey: ChunkNotFound,
+                     client.exceptions.NoSuchBucket: ChunkNotFound}
+        super(S3ChunkStore, self).__init__(error_map)
+        self.client = client
+
+    @classmethod
+    def from_url(cls, url):
+        """Construct S3 chunk store from endpoint URL.
+
+        S3 authentication (i.e. the access + secret keys) is handled externally
+        via the botocore config file or environment variables.
+
+        Parameters
+        ----------
+        url : string
+            Endpoint of S3 service, e.g. 'http://127.0.0.1:9000'
+
+        Raises
+        ------
+        ImportError
+            If botocore is not installed (it's an optional dependency otherwise)
+        :exc:`chunkstore.StoreUnavailable`
+            If S3 server interaction failed (it's down, no authentication, etc)
+        """
+        if not botocore:
+            raise ImportError('Please install botocore for katdal S3 support')
+        session = botocore.session.get_session()
+        config = botocore.config.Config(max_pool_connections=200,
+                                        s3={'addressing_style': 'path'})
+        error_map = {EndpointConnectionError: StoreUnavailable,
+                     NoCredentialsError: StoreUnavailable,
+                     ValueError: StoreUnavailable}
+        with ChunkStore(error_map)._standard_errors():
+            client = session.create_client(service_name='s3',
+                                           endpoint_url=url, config=config)
+            # Quick smoke test to see if the S3 server is available
+            client.list_buckets()
+        return cls(client)
+
+    def get(self, array_name, slices, dtype):
+        """See the docstring of :meth:`ChunkStore.get`."""
+        chunk_name, shape = self.chunk_metadata(array_name, slices, dtype=dtype)
+        bucket, key = self.split(chunk_name, 1)
+        with self._standard_errors(chunk_name):
+            response = self.client.get_object(Bucket=bucket, Key=key)
+        with contextlib.closing(response['Body']) as stream:
+            data_str = stream.read()
```

Comment:
`S3ChunkStore` depends on `botocore` which depends on its own version of `requests` which depends on `urllib3` which resembles `httplib`. The `stream.read` call passes on to a "raw" HTTP response file-like object and apparently resembles `httplib.HTTPResponse.read`. It therefore might raise something like `httplib.IncompleteRead`, although the general approach to streaming HTTP content seems to be to bumble on and let the user figure out any missing data. The `BadChunk` check after this line should detect it then.

Short answer: If you can tell me how to test it in a unit test, I'll handle it. Otherwise, I'm ignoring it for now, until a real exception happens in the wild to clue me in. Then I'll move the second `with` block inside the first `with` block and add the relevant error handling.

Reply:
Urgh, you're right, it's a real onion and no way to tell from the documentation what will happen. I think the only way to get a representative unit test would be to write some sort of proxy between the test and fakes3 that would allow for fault injection. That sounds like far too much work at this point.